limitador/src/storage/redis/redis_lenient.rs

```diff
@@ -0,0 +1,155 @@
+use crate::prometheus_metrics::CounterAccess;
+use crate::storage::in_memory::InMemoryStorage;
+use crate::storage::redis::AsyncRedisStorage;
+use crate::storage::{AsyncCounterStorage, Authorization, CounterStorage, StorageErr};
+use async_trait::async_trait;
+use std::collections::HashSet;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::time::Duration;
+
+pub struct RedisLenient {
+    redis: AsyncRedisStorage,
+    fallback: InMemoryStorage,
+    partitioned: AtomicBool,
+}
+
+#[async_trait]
+impl AsyncCounterStorage for RedisLenient {
+    async fn is_within_limits(&self, counter: &Counter, delta: i64) -> Result<bool, StorageErr> {
+        if self.is_partitioned() {
+            self.fallback.is_within_limits(counter, delta)
+        } else {
+            self.redis
+                .is_within_limits(counter, delta)
+                .await
+                .or_else(|err| {
+                    if err.is_transient() {
+                        self.partitioned(true);
+                        self.fallback.is_within_limits(counter, delta)
+                    } else {
+                        Err(err)
+                    }
+                })
+        }
+    }
+
+    async fn update_counter(&self, counter: &Counter, delta: i64) -> Result<(), StorageErr> {
+        if self.is_partitioned() {
+            self.fallback.update_counter(counter, delta)
+        } else {
+            self.redis
+                .update_counter(counter, delta)
+                .await
+                .or_else(|err| {
+                    if err.is_transient() {
+                        self.partitioned(true);
+                        self.fallback.update_counter(counter, delta)
+                    } else {
+                        Err(err)
+                    }
+                })
+        }
+    }
+
+    async fn check_and_update<'a>(
+        &self,
+        counters: &mut Vec<Counter>,
+        delta: i64,
+        load_counters: bool,
+        counter_access: CounterAccess<'a>,
+    ) -> Result<Authorization, StorageErr> {
+        if self.is_partitioned() {
+            self.fallback
+                .check_and_update(counters, delta, load_counters)
+        } else {
+            self.redis
+                .check_and_update(counters, delta, load_counters, counter_access)
+                .await
+                .or_else(|err| {
+                    if err.is_transient() {
+                        self.partitioned(true);
+                        self.fallback
+                            .check_and_update(counters, delta, load_counters)
+                    } else {
+                        Err(err)
+                    }
+                })
+        }
+    }
+
+    async fn get_counters(&self, limits: HashSet<Limit>) -> Result<HashSet<Counter>, StorageErr> {
+        if self.is_partitioned() {
+            self.fallback.get_counters(&limits)
+        } else {
+            self.redis
+                .get_counters(limits.clone())
```

Comment:
Refactor this to have the `limits: HashSet<Limit>` be returned back as part of the `Err` to avoid the (early?) `.clone()` here